Support dynamic no. of assets to support multiple project types

tnagorra opened the pull request:
Depends on
- https://github.com/mapswipe/mapswipe-backend/pull/14

## Changes

- Asset Management
  - Add `ProjectAsset` model
  - Remove `aoi_geometry_file`, `processed_geometry_file` from project
  - Add `project_type_specific_output` field to store output file
  - Add mutations
  - Add queries
- Project Create/Update
  - Separate out serializers and inputs for `ProjectCreate` and `ProjectUpdate`
  - Add default values for mandatory fields not included in project create
  - Handle case for undefined `project_type_specifics` on update
- Refactor
  - Move geojson creation from base class to tile map service class
  - Use `pk` instead of `id`
  - Configure project assets directory

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] n+1 queries
- [x] flake8 issues
- [x] `print`
- [x] typos
- [x] unwanted comments

## This PR contains valid:

- [x] tests
- [x] permission checks (tests here too)

tnagorra committed:
Add ProjectAsset model

- Remove aoi_geometry_file, processed_geometry_file from Project
- Add project_type_specific_output field to store output file

tnagorra committed:
Remove fields from Project Create mutation

- Separate out serializers/inputs for ProjectCreate and ProjectUpdate
- Add default values for mandatory fields not included in project create
- Handle undefined project_type_specifics on update
- Remove requesting_organization subfactory
- Remove fields from Project factory with default values
- Update tests

tnagorra committed:
Move geojson creation from base class to tile map service class

tnagorra committed:
Add mutation to create project asset

- Update project asset type
- Create and attach asset on project creation
- Add aoiGeometry (str) on project specifics
- Fetch/retrive aoiGeoemtry from project specifics and assets
- Fix issue with project type specific output
- Pass user on project process task

tnagorra committed:
Add migration file and update schema

tnagorra committed:
Add typehints for all models

- Use 'pk' instead of 'id'
- Use Upload.project_asset to get project assets directory
- Add comments for strawberry config (for schema generation)

tnagorra committed:
Add queries for project assets

tnagorra committed:
Rename "MimeType" to "Mimetype"

thenav56 commented on README.md:
````diff
@@ -65,9 +65,9 @@ docker compose exec web ./manage.py changepassword mapswipe-user-1
 
 ```bash
-docker compose run --rm web ./manage.py graphql_schema --out schema.graphql
+docker compose run --rm web ./manage.py export_schema main.graphql.schema --path schema.graphql
````
Let's use the custom one

thenav56 commented on apps/contributor/models.py:
```diff
@@ -37,6 +37,9 @@ class ContributorUserGroup(UserResource):
     )
 
+    # Type hints
+    archived_by_id: int
```
Let's update the archived_by to nullable

thenav56 commented on apps/project/graphql/types.py:
```diff
@@ -51,10 +51,18 @@ class ProjectType:
     max_tasks_per_user: strawberry.auto
-    project_type_specifics: CompareProjectPropertyType | FindProjectPropertyType
-    # TODO(tnagorra): Add aoi_geometry_file
+    # FIXME(tnagorra): Need to verify if we need to also add "None" here
```
```suggestion
```

thenav56 commented on apps/project/graphql/types.py:
```diff
@@ -51,10 +51,18 @@ class ProjectType:
+    type: strawberry.auto
+    mimetype: strawberry.auto
+    project_id: strawberry.ID
+    # TODO(tnagorra): Do we also add project here?
```
```suggestion
```

thenav56 commented on apps/project/models.py:
```diff
@@ -13,6 +13,32 @@
+
+
+class ProjectAssetTypeEnum(models.IntegerChoices):
+    INPUT = 100, "Input"
```
Let's add FIXME for name finalizing

thenav56 commented on apps/project/models.py:
```diff
@@ -197,12 +226,21 @@ class Project(UserResource):
-    project_type_specifics = models.JSONField(blank=True)
+    project_type_specifics = models.JSONField(blank=True, null=True)
+
+    project_type_specific_output: "ProjectAsset" = models.ForeignKey(  # type: ignore[reportAssignmentType]
```
```suggestion
    project_type_specific_output: "ProjectAsset | None" = models.ForeignKey(  # type: ignore[reportAssignmentType]
```

thenav56 commented on apps/project/models.py:
```diff
@@ -282,6 +304,34 @@ def clean(self):
+    )
+
+    file = models.FileField(
+        # FIXME(tnaogrra): the upload_to might not be correct
```
```suggestion
```

thenav56 commented on apps/project/project_types/base/project.py:
```diff
@@ -47,9 +44,10 @@ class BaseProject(
     project_task_property_class: type[ProjectTaskPropertyTypeVar]
 
-    def __init__(self, project: Project):
+    def __init__(self, project: Project, user: User):
```
Let's use self.project.modified_by

thenav56 commented on apps/project/project_types/tile_map_service/base/project.py:
```diff
@@ -20,6 +32,7 @@
 class TileMapServiceProjectProperty(base_project.BaseProjectProperty):
     zoom_level: int
     tile_server_property: TileServerConfig
+    aoi_geometry: str  # NOTE: Should be numeric
```
Let's use a custom pydantic field to validate the input (check database, check if it's a valid number)
https://docs.pydantic.dev/latest/concepts/fields/

thenav56 commented on apps/project/project_types/tile_map_service/base/project.py:
```diff
@@ -69,10 +82,68 @@ class TileMapServiceBaseProject(
+            type=ProjectAssetTypeEnum.OUTPUT,
+            mimetype=ProjectAssetMimetypeEnum.GEOJSON,
+            # FIXME(tnagorra): Had to pass user just to set these 2 fields
+            # Need to check if we can skip this
```
Let's add another FIXME for maybe using an internal user like mapswipe-bot?

thenav56 commented on apps/project/serializers.py:
```diff
@@ -38,9 +38,30 @@
+            "look_for",
+            "additional_info_url",
+            "description",
+            "image",
```
Let's move the image to update serializer

tnagorra replied:
We will have to make `image` non-mandatory from the database.
Do we need to add required validation when updating the project as "Marked as ready"?

thenav56 commented on apps/user/models.py:
```diff
@@ -17,7 +17,7 @@ class User(AbstractUser):
     objects: CustomUserManager = CustomUserManager()  # type: ignore[reportAssignmentType]
 
-    pk: int
+    # pk: int
```
```suggestion
```

tnagorra committed:
Add FIXME comments

tnagorra committed:
amend! Add typehints for all models

Add typehints for all models

- Use 'pk' instead of 'id'
- Use Upload.project_asset to get project assets directory
- Add comments for strawberry config (for schema generation)

tnagorra committed:
Use project modified_by to create project asset

tnagorra committed:
amend! Add typehints for all models

Add typehints for all models

- Use 'pk' instead of 'id'
- Use Upload.project_asset to get project assets directory
- Add comments for strawberry config (for schema generation)

tnagorra committed:
Implement pydnatic field validators

- Add model validator to check aoi_geometry

susilnem commented on apps/contributor/models.py:
```diff
@@ -37,6 +37,9 @@ class ContributorUserGroup(UserResource):
         on_delete=models.PROTECT,
```
```suggestion
        on_delete=models.SET_NULL
        blank=True,
        null=True,
```

tnagorra committed:
Make archived_by nullable field
